sycl/source/detail/error_handling/enqueue_kernel.cpp

```diff
@@ -170,13 +170,52 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
       "OpenCL API failed. OpenCL API returns: " + codeToString(Error), Error);
 }
 
+bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
+                                const NDRDescT &NDRDesc) {
+  const bool HasLocalSize = (NDRDesc.LocalSize[0] != 0);
+
+  const plugin &Plugin = DeviceImpl.getPlugin();
+  RT::PiDevice Device = DeviceImpl.getHandleRef();
+
+  if (HasLocalSize) {
+    size_t maxThreadsPerBlock[3] = {};
+    Plugin.call<PiApiKind::piDeviceGetInfo>(
+        Device, PI_DEVICE_INFO_MAX_WORK_ITEM_SIZES, sizeof(maxThreadsPerBlock),
+        maxThreadsPerBlock, nullptr);
+
+    for (size_t i = 0; i < 3; ++i) {
+      if (maxThreadsPerBlock[i] < NDRDesc.LocalSize[i]) {
+        throw sycl::nd_range_error(
+            "The number of work-items in each dimension of a work-group cannot "
+            "exceed info::device::max_work_item_sizes which is {" +
+                std::to_string(maxThreadsPerBlock[0]) + ", " +
+                std::to_string(maxThreadsPerBlock[1]) + ", " +
+                std::to_string(maxThreadsPerBlock[2]) + "} for this device",
+            PI_INVALID_WORK_GROUP_SIZE);
+      }
+    }
+  }
+
+  // Fallback
+  constexpr pi_result Error = PI_INVALID_WORK_GROUP_SIZE;
+  throw runtime_error(
+      "PI backend failed. PI backend returns: " + codeToString(Error), Error);
+}
+
 bool handleError(pi_result Error, const device_impl &DeviceImpl,
                  pi_kernel Kernel, const NDRDescT &NDRDesc) {
   assert(Error != PI_SUCCESS &&
          "Success is expected to be handled on caller side");
   switch (Error) {
-  case PI_INVALID_WORK_GROUP_SIZE:
+  case PI_INVALID_WORK_GROUP_SIZE: {
+    std::string PlatformName =
+        DeviceImpl.get_platform().get_info<info::platform::name>();
+    // TODO: Find a better way to determine the backend
```

Comment:
I think this is a good way. I am not sure if it is mandatory for this information to be a part of platform name. Do you know?
Another way will be to check the SYCL_BE env variable, but i will suggest to stick to this. We would like to support simultaneous connection to multiple backends in the future, so checking supported backend for each device/platform looks good.
Maybe sycl will develop its own extensions list and enable something like device.has_extension("OpenCL") API usage.

Reply:
Alternatively plugins could carry some simple information about the BE it represents, which could then be queried.

Reply:
Yes. We do not have any such thing for now. Hopefully we will have something soon.